### PR TITLE
warnings or errors with a known fix

### PR DESCRIPTION
**Pull request by gares**
This PR adds warnings and errors with a quickfix payload, and implements it for some easy cases:
- `intros.`
- `Extraction ambiguous.name.`
- `Axioms only_one : axiom.`
- `Module A. End B.`

Overlay PR:
- https://github.com/coq-community/vscoq/pull/793
- https://github.com/ejgallego/coq-lsp/pull/795
- https://github.com/uds-psl/autosubst-ocaml/pull/18
- https://github.com/coq-tactician/coq-tactician/pull/85
- https://github.com/ejgallego/coq-serapi/pull/421

### CODE REVIEW

File: lib/cWarnings.ml [hunk 1]
```diff
@@ -108,12 +108,12 @@ let () = CErrors.register_handler (function
     | WarnError w -> Some (print w)
     | _ -> None)
 
-let warn { Msg.tag; warning } ?loc v =
+let warn { Msg.tag; warning } ?loc ?quickfix v =
   let tag = DMap.tag_of_onetag tag in
   match warning_status warning with
   | Disabled -> ()
   | AsError -> Loc.raise ?loc (WarnError (W (tag, v)))
```
Comment (SkySkimmer) [inline lib/cWarnings.ml:115]: Is it intended that AsError warnings don't have the quickfix?

File: plugins/ltac/tacinterp.ml [hunk 2]
```diff
@@ -2017,6 +2017,30 @@ end
...
+      let new_hyps = Proofview.Goal.hyps gl in
+      let old_ids = List.map Context.Named.Declaration.get_id old_hyps in
+      let new_ids = List.map Context.Named.Declaration.get_id new_hyps in
+      let nnew = List.length new_ids in
+      let nold = List.length old_ids in
+      let names, rest = CList.chop (nnew - nold) new_ids in
+      let names = List.rev names in
+      if names <> [] && Option.has_some loc then
+        warning_fragile_intros ?loc
+          ~quickfix:[Feedback.Quickfix.make ~loc:Option.(get loc) Pp.(str "intros " ++ prlist_with_sep spc Id.print names)] ();
```
Comment (SkySkimmer) [inline plugins/ltac/tacinterp.ml:2038]: Will this behave acceptably if I have `split; intros` where the executions of `intros` produce different names?

Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml:2038]: >In the code below I think I call this only if the toplevel tactic is intros.

I don't think that's correct.

If you have something that uses the quickfix you can test with
~~~coq
Goal (forall x:nat, x = x) /\ (forall y:bool, y = y).
  split; intros.
~~~

Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml:2038]: also worth considering
~~~coq
Ltac the_tac := split; intros; reflexivity.

Goal (forall x:nat, x = x) /\ (forall y:bool, y = y).
  split; intros.
~~~

File: lib/cWarnings.ml (hunk 1 above)
Comment (rtetley) [reply lib/cWarnings.ml:115]: Indeed they can

File: plugins/ltac/tacinterp.ml (hunk 2 above)
Comment (rtetley) [reply plugins/ltac/tacinterp.ml:2038]: After testing,
```coq 
Goal (forall x:nat, x = x) /\ (forall y:bool, y = y).
  split; intros.
```
does not seem to trigger a warning

Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml:2038]: >I don't think that's correct.

I was wrong.

Comment (gares) [reply plugins/ltac/tacinterp.ml:2038]: ok, so I consider this discussion closed

File: lib/cWarnings.ml (hunk 1 above)
Comment (gares) [reply lib/cWarnings.ml:115]: I'm swamped this week.  I guess the fix is about putting the quickfix argument also in `W`, @rtetley if you have a minute, help is welcome.

Comment (rtetley) [reply lib/cWarnings.ml:115]: I'll look into it, but most likely not before Wednesday

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@SkySkimmer would you assign?

File: doc/sphinx/proof-engine/tactics.rst [hunk 3]
```diff
@@ -1241,6 +1241,17 @@ Managing the local context
 
          There are fewer than :n:`@natural` premises in the goal.
 
+      .. warn:: intros is fragile, specify the names.
+
+         When no name is specified the tactic generates fresh names depending on
+         the current context, and these names can be used in the fillowing
```
Comment (proux01) [inline doc/sphinx/proof-engine/tactics.rst:1247]:
```suggestion
         the current context, and these names can be used in the following
```

File: doc/sphinx/proof-engine/tactics.rst [hunk 4]
```diff
@@ -1241,6 +1241,17 @@ Managing the local context
...
+         proof text. Moving the proof script to another file may change the
```
Comment (proux01) [inline doc/sphinx/proof-engine/tactics.rst:1248]:
```suggestion
         proof text. This is brittle as any change may not preserve the
```

**Review by proux01 — APPROVED** [review]
Doc LGTM

File: lib/quickfix.ml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+type t = Loc.t * Pp.t
+let make ~loc txt = loc, txt
+let pp (_,x) = x
+let loc (l,_) = l
+let quickfix : t list Exninfo.t = Exninfo.make "quickfix"
+
+let raise ?loc ?qf e =
+  match qf with
+  | None     -> Loc.raise ?loc:loc e
```
Comment (SkySkimmer) [inline lib/quickfix.ml:21]:
```suggestion
  | None -> Loc.raise ?loc:loc e
```

File: lib/system.ml [hunk 6]
```diff
@@ -90,6 +90,7 @@ let process_subdirectories f path =
 let warn_cannot_open_dir =
   CWarnings.create ~name:"cannot-open-dir" ~category:CWarnings.CoreCategories.filesystem
   (fun dir -> str ("Cannot open directory " ^ dir))
+let warn_cannot_open_dir ?loc d = warn_cannot_open_dir ?loc d
```
Comment (SkySkimmer) [inline lib/system.ml:93]: ??

Comment (SkySkimmer) [inline lib/system.ml:93]: It looks annoying to have to do this everywhere a warning is exposed, how about instead leave the type of `CWarnings.create` as before and add a `CWarning.create_with_qf` that has the new type?

File: vernac/g_vernac.mlg [hunk 7]
```diff
@@ -197,6 +197,12 @@ END
 let warn_plural_command =
   CWarnings.create ~name:"plural-command" ~category:CWarnings.CoreCategories.pedantic ~default:CWarnings.Disabled
          (fun kwd -> strbrk (Printf.sprintf "Command \"%s\" expects more than one assumption." kwd))
+let warn_plural_command ?loc k =
+  match loc with
+  | None -> warn_plural_command k
+  | Some loc ->
+      let kwd = Str.replace_first (Str.regexp "s$") "" k in
```
Comment (SkySkimmer) [inline vernac/g_vernac.mlg:204]: that's not correct is it? on `Hypotheses` it will produce `Hypothese` instead of `Hypothesis`

Comment (gares) [reply vernac/g_vernac.mlg:204]: good catch! I'll remove this last commit from the PR, there is also a TODO with a wrong loc that I don't have the time to fix

File: lib/system.ml (hunk 6 above)
Comment (gares) [reply lib/system.ml:93]: ok, on it

File: plugins/ltac/tacinterp.ml [hunk 8]
```diff
@@ -2043,9 +2067,10 @@ let hide_interp {global;ast} =
   let hide_interp env =
     let ist = Genintern.empty_glob_sign ~strict:false env in
     let te = intern_pure_tactic ist ast in
-    let t = eval_tactic te in
-    t
-  in
+    match ast.CAst.v with
+    | TacAtom (TacIntroPattern(false,[ { CAst.v = IntroForthcoming false; loc } ])) ->
+      intro_patterns_deprecated false [CAst.make ?loc @@ IntroForthcoming false] te.CAst.loc
```
Comment (SkySkimmer) [inline plugins/ltac/tacinterp.ml:2072]: I don't really like doing this here, how about instead passing the info of being at toplevel down to the interp functions? something like
~~~
modified   plugins/ltac/tacinterp.ml
@@ -1113,14 +1113,14 @@ let rec val_interp ist ?(appl=UnnamedAppl) (tac:glob_tactic_expr) : Val.t Ftacti
   | _ -> value_interp ist >>= fun v -> return (name_vfun appl v)
 
 
-and eval_tactic_ist ist tac : unit Proofview.tactic =
+and eval_tactic_ist ?toplevel ist tac : unit Proofview.tactic =
   let (loc, tac2) = CAst.(tac.loc, tac.v) in
   match tac2 with
   | TacAtom t ->
       let call = LtacAtomCall t in
       let (stack, _) = push_trace(loc,call) ist in
       do_profile stack
-        (catch_error_tac_loc loc stack (interp_atomic ist t))
+        (catch_error_tac_loc loc stack (interp_atomic ?toplevel ist t))
   | TacFun _ | TacLetIn _ | TacMatchGoal _ | TacMatch _ -> interp_tactic ist tac
   | TacId [] -> Proofview.tclLIFT (db_breakpoint (curr_debug ist) [])
   | TacId s ->
@@ -1666,9 +1666,11 @@ and name_atomic ?env tacexpr tac : unit Proofview.tactic =
   Proofview.Trace.name_tactic name tac
 
 (* Interprets a primitive tactic *)
-and interp_atomic ist tac : unit Proofview.tactic =
+and interp_atomic ?toplevel ist tac : unit Proofview.tactic =
   match tac with
   (* Basic tactics *)
+  | TacIntroPattern(false,[ { CAst.v = IntroForthcoming false; loc } ]) when Option.has_some toplevel ->
+    intro_patterns_deprecated false [CAst.make ?loc @@ IntroForthcoming false] toplevel
   | TacIntroPattern (ev,l) ->
       Proofview.Goal.enter begin fun gl ->
         let env = Proofview.Goal.env gl in
@@ -1968,14 +1970,14 @@ let default_ist () =
   let extra = TacStore.set TacStore.empty f_debug (get_debug ()) in
   { lfun = Id.Map.empty; poly = false; extra = extra }
 
-let eval_tactic t =
+let eval_tactic ?toplevel t =
   if get_debug () <> DebugOff then
     Proofview.tclUNIT () >>= fun () -> (* delay for [default_ist] *)
     Proofview.tclLIFT (db_initialize true) <*>
-    eval_tactic_ist (default_ist ()) t
+    eval_tactic_ist ?toplevel (default_ist ()) t
   else
     Proofview.tclUNIT () >>= fun () -> (* delay for [default_ist] *)
-    eval_tactic_ist (default_ist ()) t
+    eval_tactic_ist ?toplevel (default_ist ()) t
 
 let eval_tactic_ist ist t =
   Proofview.tclLIFT (db_initialize false) <*>
@@ -2043,7 +2045,7 @@ let hide_interp {global;ast} =
   let hide_interp env =
     let ist = Genintern.empty_glob_sign ~strict:false env in
     let te = intern_pure_tactic ist ast in
-    let t = eval_tactic te in
+    let t = eval_tactic ?toplevel:te.loc te in
     t
   in
   if global then
~~~

Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml:2072]: It may also be that this warning isn't worth implementing

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
there should be tests for each added quickfix

File: lib/cErrors.mli [hunk 9]
```diff
@@ -34,7 +34,7 @@ exception UserError of Pp.t
 (** Main error signaling exception. It carries a header plus a pretty printing
     doc *)
 
-val user_err : ?loc:Loc.t -> ?info:Exninfo.info -> Pp.t -> 'a
+val user_err : ?loc:Loc.t -> ?info:Exninfo.info -> ?qf:Quickfix.t list -> Pp.t -> 'a
```
Comment (SkySkimmer) [inline lib/cErrors.mli:37]: What happens to the quickfix data? It doesn't seem like anything reads it.

**Review by ejgallego — COMMENTED** [review]
Indeed we should be careful about the model here, often Quicfix data is very expensive to compute (that's why the LSP protocol actually use `resolve` for it).

So while this model can be OK "as a prototype", performance should be tested for each QuickFix until the data can be computed on demand.

Maybe we should indeed wait and do the right thing, that is to say, postpone the computation of QF data to the vernac layer so language manager can properly implement the standard performance profile.

What do you think @SkySkimmer ?

**Comment by ejgallego** [comment]
I've added a `needs: discussion` tag based on my review.

File: plugins/ltac/tacinterp.ml (hunk 8 above)
Comment (ejgallego) [reply plugins/ltac/tacinterp.ml:2073]: > I don't really like doing this here, how about instead passing the info of being at toplevel down to the interp functions? something like

Indeed, this model while simple, can come back to bite us in the end. Maybe we should have an exception / warning wrapper that can capture the needed closure data, and let the upper layer compute the quickfix too.

This has several advantates:
- matches the lsp performance profile
- is modular and allow people to implement their own QF plugins.

File: lib/cErrors.mli (hunk 9 above)
Comment (gares) [reply lib/cErrors.mli:37]: done

File: plugins/ltac/tacinterp.ml (hunk 8 above)
Comment (gares) [reply plugins/ltac/tacinterp.ml:2073]: I think it is, although not necessarily in this PR

**Comment by gares** [comment]
@coqbot run full ci

File: lib/cWarnings.mli [hunk 10]
```diff
@@ -56,6 +56,10 @@ val create : name:string -> ?category:category -> ?default:status ->
   ('a -> Pp.t) -> ?loc:Loc.t -> 'a -> unit
 (** Combined creation function. [name] must be a fresh name. *)
 
+val create_with_quickfix : name:string -> ?category:category -> ?default:status ->
+    ('a -> Pp.t) -> ?loc:Loc.t -> ?quickfix:Quickfix.t list -> 'a -> unit
```
Comment (SkySkimmer) [inline lib/cWarnings.mli:60]: An alternative API could be something like
```suggestion
val create : name:string -> ?category:category -> ?default:status ->
  ?quickfix:('a -> Quickfix.t list) ->
  ('a -> Pp.t) -> ?loc:Loc.t -> 'a -> unit
```
(since the quickfix generator is optional we don't need a separate create_with_quickfix)

**Comment by gares** [comment]
In this PR I add quickfix for `intros` that suggests the names actually used. That could have a non negligible cost.

As pointed out above the change is not very nice, since recognizing a "toplevel" intros is not immediate. So I will remove it from this PR anyway, but before I'll do the bench anyway.

Other than that, what are the expensive quickfixes you have in mind @ejgallego ?

**Comment by gares** [comment]
@cobot bench

**Comment by gares** [comment]
@coqbot bench

File: lib/cWarnings.mli (hunk 10 above)
Comment (gares) [reply lib/cWarnings.mli:60]: Sorry @SkySkimmer I though you wanted this to avoid some spurious lines in the patch that were eliding the optional argument after declaring the warning. Did I get it wrong?

File: lib/cErrors.ml [hunk 11]
```diff
@@ -126,7 +133,7 @@ let print_gen ~anomaly (e, info) =
 
 (** The standard exception printer *)
 let iprint (e, info) =
-  print_gen ~anomaly:true (e,info) ++ print_backtrace info
+  print_gen ~anomaly:true (e,info) ++ print_quickfix info ++ print_backtrace info
```
Comment (SkySkimmer) [inline lib/cErrors.ml:136]: should use register_additional_error_info instead of additional adhoc handling

**Comment by SkySkimmer** [comment]
>In this PR I add quickfix for intros that suggests the names actually used. That could have a non negligible cost.

I don't think it's likely to be expensive (we have to generate the names to intro them anyway), I just prefer to avoid splitting the implementation of intros to relatively distant parts of the code.

File: lib/cErrors.ml (hunk 11 above)
Comment (gares) [reply lib/cErrors.ml:136]: Thanks for the hint, I did not know that API existed

Comment (SkySkimmer) [reply lib/cErrors.ml:136]: also I don't know if using exninfo is the right way to do it, maybe a table of `exn -> quickfix option` functions would be better
then in cwarnings we would have `type w = W : 'a tag * quickfix option * 'a -> w` instead of the current `type w = W : 'a tag * 'a -> w`, or maybe delay the `'a -> quickfix` computation as is done for printing

Comment (ejgallego) [reply lib/cErrors.ml:136]: Why not a `('a, 'qc) With_quickfix_info.t = { 'a; qc_info : unit -> 'qc }` wrapper, that would allow the upper layers to apply the quickfixes later? I dunno, there are several designs here.

**Comment by ejgallego** [comment]
> Other than that, what are the expensive quickfixes you have in mind @ejgallego ?

There are many examples, think of a missing import for example, it could be pretty expensive to suggest what the right file would be, many kind of proof repairs, etc... Actually I have problem coming with Quickfixes that are not expensive.

Anyways the LSP spec works this way, so it is good we try to stay close to it.

In the long term I think it is important that:
- the QC can be computed in a delayed way (not for all, if a QC is trivial)
- the QC are easily extensible without messing with a lot of internals

**Comment by ejgallego** [comment]
> In this PR I add quickfix for intros that suggests the names actually used. That could have a non negligible cost.

Actually it would have been very interesting to see the bench with this PR added, to get an idea if they do have any real cost.

**Comment by gares** [comment]
> Actually it would have been very interesting to see the bench with this PR added, to get an idea if they do have any real cost.

As I write above, the bench includes the commit for `intros`. Not a performance problem after all.

File: lib/cErrors.ml (hunk 11 above)
Comment (gares) [reply lib/cErrors.ml:136]: > also I don't know if using exninfo is the right way to do it

Why not? It is a general mechanism to enrich an exception and is implemented using a table. What are the advantages of a quickfix-specific table?

File: vernac/topfmt.ml [hunk 12]
```diff
@@ -149,8 +149,10 @@ let gen_logger dbg warn ?pre_hdr level msg = let open Feedback in match level wi
   | Debug   -> msgnl_with !std_ft (make_body dbg  dbg_hdr ?pre_hdr msg)
   | Info    -> msgnl_with !std_ft (make_body dbg info_hdr ?pre_hdr msg)
   | Notice  -> msgnl_with !std_ft (make_body noq info_hdr ?pre_hdr msg)
-  | Warning -> Flags.if_warn (fun () ->
+  | Warning [] -> Flags.if_warn (fun () ->
                msgnl_with !err_ft (make_body warn warn_hdr ?pre_hdr msg)) ()
+  | Warning qf -> Flags.if_warn (fun () ->
+              msgnl_with !err_ft (make_body warn warn_hdr ?pre_hdr (msg ++ fnl () ++ str"Quickfix:" ++ fnl () ++ prlist_with_sep fnl Quickfix.pp qf))) ()
```
[the remark this answers is not in the document]
Comment (gares) [reply vernac/topfmt.ml:155]: possibly, I'll have a look.

**Comment by gares** [comment]
If you want lazily-computed quickfixed, I'd rather not use an ocaml thunk, since the state of Coq has global components, hence running the thunk later may not be trivial (unless you grab the global state in the thunk, as Futures were doing).

I think the best we can do, for now, is to add a constructor, eg `type Quickfix.t = Basic of Pp.t * Loc.t` so that when one needs to add a more sophisticated case the type checker can help him.

> think of a missing import for example

This is a kind of counter example, IMHO. The quickfix infrastructure we propose here is for things Coq can suggest how to fix. A missing dependency is something you can only fix by talking to the package manager, or something like that. In that case you want the error to be easy to recognize (i.e. not a UserError) and have the language-server to do something about it.

File: vernac/topfmt.ml (hunk 12 above)
[the remark this answers is not in the document]
Comment (gares) [reply vernac/topfmt.ml]: done

**Comment by SkySkimmer** [comment]
>I think the best we can do, for now, is to add a constructor, eg type Quickfix.t = Basic of Pp.t * Loc.t so that when one needs to add a more sophisticated case the type checker can help him.

I don't see how that's better than thunks

**Comment by gares** [comment]
> I don't see how that's better than thunks

What I mean is that a thunk in presence of a global state may not capture everything (automatically). Sure, if the code you call is pure then it works as expected.

Anyway, I think it makes much more sense to refine the type when such a use case arises.

Also, putting a thunk may not help, especially if one wants to do something with the data carried by the quickfix.
In the example of a missing import, I think the language server is much happier with an immediate `MissingImport of string` rather than a thunk generating a `Pp.t`, since it can ask `opam` where to find that import. This is why I was proposing to use `Basic of Pp.t * loc` for quickfixes that are immediately produced by Coq, and have future use cases extend the data type with whatever they may need.

File: lib/cErrors.ml (hunk 11 above)
Comment (gares) [reply lib/cErrors.ml:136]: Hum, I see what you mean by fragile, but I'm not so sure a map `exn -> whatever` can do much better. After all if one catches an exception he may as well throw another one (with or without the same info attached). I'd rather avoid 2 capture/reraise APIs.

Also note that Loc.t is the info as well, and although not critical it is quite relevant to the user.

**Comment by gares** [comment]
@SkySkimmer I did re-ask for a review sine I believe I did add the tests. So if there is anything else you consider blocking better it be in the review so that one can easily find it.

File: lib/cErrors.ml (hunk 11 above)
Comment (SkySkimmer) [reply lib/cErrors.ml:136]: >I'd rather avoid 2 capture/reraise APIs.

I am not proposing a capture API, I propose that quickfix is like printing

Comment (gares) [reply lib/cErrors.ml:136]: Ok, so you propose that each error with a quickfix should have a dedicated exception (not a UserError) and that exception should carry the quickfix. For each of such errors one registers a function to Quickfix.t, that is used by the toplevel in order to get the quickfix data out of the exception.

I think I get it now.

Comment (SkySkimmer) [reply lib/cErrors.ml:136]: Yes.
And then the function can be trivial (`MyExn (_, qf) -> qf`) or less trivial if we want to delay computation

**Comment by ejgallego** [comment]
> Probably we should schedule a coq call topic (or wait until CUDW??)

I don't know, maybe we can just discuss here? I feel that I have said most of what I had to say; I'm happy if you want to remove the tag.

I get the feelings that linters maybe should live at an upper layer?

This way people can contribute them more easily. For example, Ram had this project to make suggestions for parsing errors, in this case, our plan was to add more structure to errors, and handle that at the top-level, so we could suggest a grammar fix.

Decoupling the generation of the warning / error from the generation of the fix is standard practice in the industry.

But for some cases, this could be complicated. I guess it is hard to know without more quickfixes examples. There are many design possibilities so in my experience getting this right requires a few iterations.

File: plugins/ltac/tacinterp.ml [hunk 13]
```diff
@@ -2017,6 +2017,31 @@ end
...
+    fun () -> Proofview.Goal.enter begin fun gl ->
+      let new_hyps = Proofview.Goal.hyps gl in
+      let old_ids = List.map Context.Named.Declaration.get_id old_hyps in
+      let new_ids = List.map Context.Named.Declaration.get_id new_hyps in
+      let nnew = List.length new_ids in
+      let nold = List.length old_ids in
+      let names, rest = CList.chop (nnew - nold) new_ids in
+      let names = List.rev names in
+      if names <> [] && Option.has_some loc then
+        warning_fragile_intros ?loc
```
Comment (ppedrot) [inline plugins/ltac/tacinterp.ml:2038]: IIUC when the as-error flag is set, this would result in a backtracking error. I think this is wrong, it should be a fatal panic otherwise we're going to get extremely weird behaviours.

Comment (ppedrot) [reply plugins/ltac/tacinterp.ml:2038]: Also, intros without names is not deprecated at all. Some use cases are perfectly legitimate, notably in proof automation.

Comment (gares) [reply plugins/ltac/tacinterp.ml:2038]: > Also, intros without names is not deprecated at all. Some use cases are perfectly legitimate, notably in proof automation.

My attempt is to capture just toplevel calls to intros, that is `intros.`

Comment (gares) [reply plugins/ltac/tacinterp.ml:2038]: > IIUC when the as-error flag is set, this would result in a backtracking error.

This is a problem, but I don't get that only affects this PR. Is there no tactic out there raising a warning?

File: lib/cErrors.ml (hunk 11 above)
Comment (gares) [reply lib/cErrors.ml]: done

**Comment by gares** [comment]
@SkySkimmer I did implement your suggestion, please review. Also chat with @ppedrot ant tell me what to do with the intros commit. I'm fine dropping it or moving to another PR.

@ejgallego the new API lets one register quickfix generators in the outern layers if the exception carries enough info to actually generate the suggestion. Moreover the suggestion is generated on demand. Of course, you may still need to modify Coq if the exception does not carry enough info. Said that, I think we have all the interest in the world to put these generators inside Coq, so that all UIs can benefit.
Defining the quickfix generator far from the declaration of the exception should only be used to make proofs of concept.

**Comment by gares** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
Not sure the quickfix should be printed by default (for interfaces which don't understand it like coqc/coqtop), maybe it should be only with test-mode. Some of them are not very readable (eg the End one).

Also not sure it makes sense to have warnings put the quickfix in a separate field of the feedback thing but errors merge it with the rest of the printing.

>tell me what to do with the intros commit. I'm fine dropping it or moving to another PR.

Let's move it to another PR.

File: lib/feedback.mli [hunk 14]
```diff
@@ -15,7 +15,7 @@ type level =
   | Debug
   | Info
   | Notice
-  | Warning
+  | Warning of Quickfix.t list
   | Error
```
Comment (SkySkimmer) [inline lib/feedback.mli:19]: >Also not sure it makes sense to have warnings put the quickfix in a separate field of the feedback thing but errors merge it with the rest of the printing.

ie here

Comment (ejgallego) [reply lib/feedback.mli:19]: Yes, it'd be nice to be uniform here.

Comment (gares) [reply lib/feedback.mli:19]: ok, I'm on it.

**Comment by ejgallego** [comment]
I see the point about upper layers being able to produce the quickfixes by themselves, however if they cannot use the infra in Coq what is the point of that infra?

I get the feeling that we are still far from the "right" design here, see for example the non-uniformity between warnings and errors. This points out to IMHO a much needed rework for the feedback system, which is not easy.

On the other hand this can be a good first step towards experimenting quickfixes , I see no harm in merging it as long as we all agree this is for now an experimental, non-stable API.

IMHO if we are serious about quickfixes, we should first do a requirements analisys, in particular, we should identify all possible quickfixes. Once we have identified them, we can think what implementation changes are needed, not the other way around.

File: lib/feedback.mli (hunk 14 above)
Comment (gares) [reply lib/feedback.mli]: done

**Comment by gares** [comment]
@SkySkimmer all done, indeed with the quickfix data in the right place the changes are more smooth

File: plugins/ltac/tacinterp.ml (hunk 13 above)
Comment (gares) [reply plugins/ltac/tacinterp.ml]: Moved to another PR, namely #19224

**Comment by gares** [comment]
@coqbot run full ci

File: lib/quickfix.mli [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    forbidden.
+
+    Exceptions that are considered anomalies should not be
+    handled by Quickfix generators.
+*)
+
+val register : (exn -> t list) -> unit
+
+(** computes all the quickfixes for a given exception *)
+val get : exn -> (t list, Exninfo.iexn) result
```
Comment (SkySkimmer) [inline lib/quickfix.mli:34]:
```suggestion
val from_exception : exn -> (t list, Exninfo.iexn) result
```

File: lib/cErrors.ml [hunk 16]
```diff
@@ -124,9 +124,15 @@ let print_gen ~anomaly (e, info) =
...
 
+let print_quickfix = function
+  | [] -> mt ()
+  | qf -> fnl () ++ Quickfix.print qf
+
 (** The standard exception printer *)
 let iprint (e, info) =
-  print_gen ~anomaly:true (e,info) ++ print_backtrace info
+  print_gen ~anomaly:true (e,info) ++
+  Result.fold ~ok:print_quickfix ~error:(print_gen ~anomaly:true) (Quickfix.get e) ++
```
Comment (SkySkimmer) [inline lib/cErrors.ml:134]: shouldn't this be sent to feedback separately now?

Comment (gares) [reply lib/cErrors.ml:134]: Hum, I think the error message business is still messy. In particular all errors are turned into user erorrs:
```
let process_expr ~state loc_ast =
  try interp_vernac ~check:true ~state loc_ast
  with reraise ->
    let reraise, info = Exninfo.capture reraise in

    (* Exceptions don't carry enough state to print themselves (typically missing the nametab)
       so we need to print before resetting to an older state. See eg #16745 *)
    let reraise = UserError (CErrors.iprint (reraise, info)) in
    (* Keep just the loc in the info as it's printed separately *)
    let info = Option.cata (Loc.add_loc Exninfo.null) Exninfo.null (Loc.get_loc info) in

    ignore(Stm.edit_at ~doc:state.doc state.sid);
    Exninfo.iraise (reraise, info)
```
If I feedback an error then it may be printed twice.

Should I pass to iprint a flag or have a twin API?
In general I think printing the quickfix as part of the error message may not be a bad idea, although what I print now is surely perfectible.

File: lib/quickfix.mli (hunk 15 above)
Comment (gares) [reply lib/quickfix.mli:34]: done

**Comment by gares** [comment]
@coqbot run full ci

File: lib/cErrors.ml (hunk 16 above)
Comment (gares) [reply lib/cErrors.ml:134]: Maybe, but if the reason is to make the code above sensible, I'm not so sure.

UserError is already used elsewhere, eg Ltac, while the code above should really throw a dedicated exception (with an explicit Pp, a loc and Quickfix). I'd rather leave this cleanup to another PR.

**Comment by gares** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
needs overlays

**Comment by gares** [comment]
@coqbot run full ci

File: lib/feedback.ml [hunk 17]
```diff
@@ -42,7 +42,7 @@ type feedback_content =
   (* Extra metadata *)
   | Custom of Loc.t option * string * xml
   (* Generic messages *)
-  | Message of level * Loc.t option * Pp.t
+  | Message of level * Loc.t option * Quickfix.t list * Pp.t
```
Comment (ejgallego) [inline lib/feedback.ml:45]: This new type for `Message` seems strange, maybe it would best to introduce a new type of message `Quickfix`.

Comment (ejgallego) [reply lib/feedback.ml:45]: For instance what are clients supposed to do when a Debug message with a quickfix payload arrives?

Comment (gares) [reply lib/feedback.ml:45]: In the beginning it was just attached to Warning, then also Error and now to Message, I think @SkySkimmer suggested it. It surely makes the code simpler when a warning becomes and error. I'm also fine adding a Quickfix message, but that would then be unrelated to any other message, and that is bad, since a fix proposal with no other message attached is going to be very surprising for the user.

I think the client should handle quickfixes the same way. I don't have a use case for a Debug message suggesting a code change, but if you receive one just display it together with the proposed fix if any.

**Comment by gares** [comment]
@coqbot run full ci

File: lib/cErrors.ml [hunk 18]
```diff
@@ -128,9 +128,14 @@ let print_gen ~anomaly (e, info) =
...
 
+let print_quickfix = function
+  | [] -> mt ()
+  | qf -> fnl () ++ Quickfix.print qf
+
 (** The standard exception printer *)
 let iprint (e, info) =
-  print_gen ~anomaly:true (e,info)
+  print_gen ~anomaly:true (e,info) ++
+  Result.fold ~ok:print_quickfix ~error:(print_gen ~anomaly:true) (Quickfix.from_exception e)
```
Comment (SkySkimmer) [inline lib/cErrors.ml:138]: >UserError is already used elsewhere, eg Ltac, while the code above should really throw a dedicated exception (with an explicit Pp, a loc, a trace and Quickfix). I'd rather leave this cleanup to another PR.

OK but we should at least check test_mode here then
(I guess the test_mode ref would need to move down to lib)

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
serapi overlay seems incomplete

**Comment by gares** [comment]
Should be better now. I can't build locally, I get a CSig problem I can't really understand, so I'm patching it blind.

File: lib/cErrors.ml (hunk 18 above)
Comment (gares) [reply lib/cErrors.ml:138]: Frankly, I don't know what it is was not in Flags in the first place. Can I put it there?

Comment (SkySkimmer) [reply lib/cErrors.ml:138]: I think we generally try to avoid putting stuff in flags (it's seen by the kernel) but for this it should be OK

**Comment by SkySkimmer** [comment]
>Should be better now. I can't build locally, I get a CSig problem I can't really understand, so I'm patching it blind.

ask for help on zulip

**Comment by herbelin** [comment]
>  I get a CSig problem I can't really understand

That may be my fault. Did you rebase wrt ejgallego/coq-lsp#767?

**Comment by gares** [comment]
Ci green

File: lib/cErrors.ml (hunk 18 above)
Comment (SkySkimmer) [reply lib/cErrors.ml:138]: still need to check test_mode here

Comment (gares) [reply lib/cErrors.ml]: done

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
ci failure on coq-lsp is unrelated

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by gares** [comment]
Thanks!
